packages/@headlessui-react/src/components/popover/popover-machine.ts

```diff
@@ -0,0 +1,161 @@
+  | { type: ActionTypes.SetPanelId; panelId: string | null }
+
+let reducers: {
+  [P in ActionTypes]: (state: State, action: Extract<Actions, { type: P }>) => State
+} = {
+  [ActionTypes.OpenPopover]: (state) => {
+    if (state.popoverState === PopoverStates.Open) return state
+    return { ...state, popoverState: PopoverStates.Open, __demoMode: false }
+  },
+  [ActionTypes.ClosePopover](state) {
+    if (state.popoverState === PopoverStates.Closed) return state
+    return { ...state, popoverState: PopoverStates.Closed, __demoMode: false }
+  },
+  [ActionTypes.SetButton](state, action) {
+    if (state.button === action.button) return state
+    return { ...state, button: action.button }
+  },
+  [ActionTypes.SetButtonId](state, action) {
+    if (state.buttonId === action.buttonId) return state
+    return { ...state, buttonId: action.buttonId }
+  },
+  [ActionTypes.SetPanel](state, action) {
+    if (state.panel === action.panel) return state
+    return { ...state, panel: action.panel }
+  },
+  [ActionTypes.SetPanelId](state, action) {
+    if (state.panelId === action.panelId) return state
+    return { ...state, panelId: action.panelId }
+  },
+}
+
+export class PopoverMachine extends Machine<State, Actions> {
+  static new({ __demoMode = false } = {}) {
+    return new PopoverMachine({
+      __demoMode,
+      popoverState: __demoMode ? PopoverStates.Open : PopoverStates.Closed,
+      buttons: { current: [] },
+      button: null,
+      buttonId: null,
+      panel: null,
+      panelId: null,
+      beforePanelSentinel: { current: null },
+      afterPanelSentinel: { current: null },
+      afterButtonSentinel: { current: null },
+    })
+  }
+
+  reduce(state: Readonly<State>, action: Actions): State {
+    return match(action.type, reducers, state, action)
+  }
+
+  actions = {
+    close: () => this.send({ type: ActionTypes.ClosePopover }),
+    refocusableClose: (
+      focusableElement?: HTMLElement | { current: HTMLElement | null } | MouseEvent<HTMLElement>
+    ) => {
+      this.actions.close()
+
+      let restoreElement = (() => {
+        if (!focusableElement) return this.state.button
+        if (DOM.isHTMLElement(focusableElement)) return focusableElement
+        if ('current' in focusableElement && DOM.isHTMLElement(focusableElement.current)) {
+          return focusableElement.current
+        }
+
+        return this.state.button
+      })()
+
+      restoreElement?.focus()
+    },
+    open: () => this.send({ type: ActionTypes.OpenPopover }),
+    setButtonId: (id: string | null) => this.send({ type: ActionTypes.SetButtonId, buttonId: id }),
+    setButton: (button: HTMLElement | null) => this.send({ type: ActionTypes.SetButton, button }),
+    setPanelId: (id: string | null) => this.send({ type: ActionTypes.SetPanelId, panelId: id }),
+    setPanel: (panel: HTMLElement | null) => this.send({ type: ActionTypes.SetPanel, panel }),
+  }
+
+  selectors = {
+    isPortalled: (state: State) => {
```

Comment:
I maintain that `cancelled` is the correct spelling but seeing `ll` here just feels… weird. 🤣 (don't change it… this is just random thoughts)